src/language/patterns.md

```diff
@@ -0,0 +1,358 @@
+and extract it in parts. In other words, the pattern _destructures_ the object:
+
+```dart
+var numList = [1, 2, 3];
+// List pattern [a, b, c] destructures the three elements from numList...
+var [a, b, c] = numList;
+// ...and assigns them to new variables.
+print(a + b + c);          
+```
+
+You can nest [any kind of pattern][types] inside a destructuring pattern, 
+For example, this case pattern matches and destructures a two-element
+list whose first element is `a` or `b`:
+
+```dart
+switch (list) {
+  case ['a' || 'b', var c]: 
+  // ...
+}
+```
+
+## Places patterns can appear
+
+You can use patterns in several places in the Dart language:
+
+<a id="pattern-uses"></a>
+
+- Local variable [declarations](#variable-declaration) and [assignments](#variable-assignment)
+- [for and for-in loops][for]
+- [if-case][if] and [switch-case][switch]
+- Control flow in [collection literals][]
+
+This section describes common use cases for matching and destructuring with patterns, 
+combining various contexts, pattern types, and surrounding code constructs.
+
+### Variable declaration
+
+Anywhere Dart allows local variable declaration, like in the body of a function,
+you can also use a _pattern variable declaration_. The pattern matches against
+and destructures the value on the right of the declaration, binding it to
+new local variables:
+
+```dart
+// Declares new variables a, b, and c.
+var (a, [b, c]) = ("str", [1, 2]);
+```
+
+A pattern variable declaration must start with either `var` or `final`, followed
+by a pattern. 
+
+### Variable assignment 
+
+A _variable assignment pattern_ on the left side of an assignment destructures the
+matched object and assigns the values to _existing_ variables,
+instead of binding new ones. 
+
+One convenient use case is swapping the values of two variables without needing
+to declare a third temporary one:
+
+```dart
+var (a, b) = ('left', 'right');
+(b, a) = (a, b); // Swap.
+print('$a $b'); // Prints "right left".
+```
+
+### Switch statements and expressions
+
+Every case clause contains a pattern. This applies to [switch statements][switch]
+and [expressions][], as well as [if-case statements][if].
+
+_Case patterns_ are [refutable][];
+they allow control flow to either match and destructure the object being switched
+on, or continue execution if the object doesn't match. Case patterns can be
+[any kind of pattern][types]:
+
+```dart
+switch (obj) {
+  case 1: ...
+    // Matches if 1 == obj.
+  case >= first && <= last: ...  
+    // Matches if the value of obj is between the constant values of 'first' and 'last'.
+  case (var a, var b): ...
+    // Matches if obj is a record with two fields, then assigns the fields to 'a' and 'b'.
+  case Rect(width: var w, height: var h): ...
+    // Matches if obj is an instance of Rect, and then against the properties of Rect.
+  default: ...
+}
+```
+
+
+Values destructured by a pattern in a case become local variables,
```

Comment:
**issue:** Write in active voice per [GSG](https://developers.google.com/style/voice).
```suggestion
Values that a pattern destructures in a case become local variables.
```